src/services/postPasteImportFixes.ts

```diff
@@ -0,0 +1,187 @@
+    cancellationToken: CancellationToken,
+    changes: textChanges.ChangeTracker,
+    originalFile?: SourceFile,
+    copyLocation?: { file: string; start: { line: number; offset: number; }; end: { line: number; offset: number; }; },
+) {
+    const updatedTargetFile = host.updateTargetFile?.(targetFile.fileName, targetFile.getText(), targetFile.getText().slice(0, pastes[0].range.pos) + pastes[0].text + targetFile.getText().slice(pastes[0].range.end));
+    const statements: Statement[] = [];
+    Debug.assert(updatedTargetFile && updatedTargetFile.updatedFile && updatedTargetFile.originalProgram && updatedTargetFile.updatedProgram);
+
+    if (originalFile) {
+        addRange(statements, originalFile.statements, copyLocation?.start.line, copyLocation ? copyLocation.end.line + 1 : undefined);
+        const usage = getUsageInfo(originalFile, statements, updatedTargetFile.originalProgram.getTypeChecker(), getExistingLocals(updatedTargetFile.updatedFile, statements, updatedTargetFile.originalProgram.getTypeChecker()));
+        const importAdder = codefix.createImportAdder(updatedTargetFile.updatedFile, updatedTargetFile.updatedProgram, preferences, host);
+
+        const imports = getImportsFromKnownOriginalFile(originalFile, targetFile, updatedTargetFile.updatedProgram, importAdder, usage.oldImportsNeededByTargetFile, usage.targetFileImportsFromOldFile, changes, preferences, host, updatedTargetFile.originalProgram.getTypeChecker());
+        if (imports.length > 0) {
+            insertImports(changes, targetFile, imports, /*blankLineBetween*/ true, preferences);
+        }
+        importAdder.writeFixes(changes, getQuotePreference(originalFile, preferences));
+        host.revertUpdatedFile?.(targetFile.fileName, updatedTargetFile.updatedFile.text, targetFile.text);
+    }
+    else {
+        const context: CodeFixContextBase = {
+            sourceFile: updatedTargetFile.updatedFile,
+            program: updatedTargetFile.originalProgram,
+            cancellationToken,
+            host,
+            preferences,
+            formatContext,
+        };
+        const importAdder = codefix.createImportAdder(updatedTargetFile.updatedFile, updatedTargetFile.updatedProgram, preferences, host);
+        forEachChild(updatedTargetFile.updatedFile, function cb(node) {
+            if (isIdentifier(node)) {
+                if (!updatedTargetFile.originalProgram?.getTypeChecker().resolveName(node.text, node, SymbolFlags.All, /*excludeGlobals*/ false)) {
+                    // generate imports
+                    importAdder.addImportsForUnknownSymbols(context, node, /*useAutoImportProvider*/ true);
+                }
+            }
+            else {
+                node.forEachChild(cb);
+            }
+        });
+        importAdder.writeFixes(changes, getQuotePreference(targetFile, preferences));
+        host.revertUpdatedFile?.(targetFile.fileName, updatedTargetFile.updatedFile.text, targetFile.text);

```

Comment:
I’m concerned about an unexpected error getting thrown between `updateTargetFile` and `revertUpdatedFile`, leaving the server in a corrupted state for subsequent commands. I think the code needs to be structured very carefully like

```ts
// Comment explaining what's about to happen
updateFile()
try {
  doEverything()
} finally {
  revertFile()
}
```

By that I primarily mean you need a try/finally, but I also think it may be worth a little aesthetic refactoring to make the `try` block very short like my pseudocode example, so that this whole dangerous sequence can easily be digested by someone reading the code in the future.